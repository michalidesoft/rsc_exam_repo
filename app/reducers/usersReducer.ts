import { USERS_ACTION_TYPES } from '../enums';
import { Action } from '../interfaces/action.interface';
import { User } from '../interfaces/user.interface';

export interface State {
    list: undefined | null | User[];
    loading: boolean;
    error: null | string;
}

export const initialState: State = {
    list: null,
    loading: false,
    error: null
};

export const usersReducer = (state = initialState, action: Action): State => {
    switch(action.type) {
        case USERS_ACTION_TYPES.USERS_LOAD:
            return {
                ...state,
                loading: true,
                error: null
            };
        case USERS_ACTION_TYPES.USERS_LOAD_ERROR: 
            return {
                ...state,
                loading: false,
                error: 'Error loading list of users',
                list: null
            }
        case USERS_ACTION_TYPES.USERS_LOAD_SUCCESS:
            return {
                ...state,
                loading: false,
                error: null,
                list: action.payload
            }
        default:
            return state;
    }
}

export default usersReducer;