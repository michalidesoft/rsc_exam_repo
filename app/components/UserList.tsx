import React, { useEffect } from 'react';
import {
  View,
  Text,
  Image,
  FlatList
} from 'react-native';
import { useSelector, useDispatch } from 'react-redux';

import fetchUsers from '../actions/fetchUsers';
import { User } from '../interfaces/user.interface';
import { State } from '../reducers/usersReducer';

/**
 * Renders the actual FlatList of users, should only be used, if props.users
 * has a length > 0
 * @param props {{ users: User[] }} list of users in in the 'users' property 
 */
const List = (props: { users: User[] }) => {
    const dimension = 60;

    return (
        <FlatList
            data={props.users}
            renderItem={({ item }) => (
                <View style={{ padding: 10, margin: 3, borderRadius: 5, backgroundColor: 'white', flexDirection: 'row' }}>
                    <Image source={{ uri: item.avatar }} style={{ width: dimension, height: dimension, borderRadius: dimension / 2 }}/>
                    <Text style={{ padding: 22 }}>{item.name}</Text>
                </View>)}
        />
    )
}


/**
 * Selects the users Redux store and renders information based on the loading state 
 * of the API call: shows a generic error message, th loading state (while the
 * call is pending) or the List of Users. Shows a message, if the list of users
 * is empty.
 */
const UserList = () => {

    const users = useSelector((state: { users: State}) => state.users);
    const dispatch = useDispatch(); 
    
    useEffect(() => {
        dispatch (fetchUsers());
    }, []);

    if (users.error) {
        return (<View><Text>Error loading users.</Text></View>)
    }

    if (users.loading) {
        return (<View><Text>Loading...</Text></View>);
    }

    if (!users.list) {
        return (<View><Text>User list is empty.</Text></View>)
    }

    return (
        <View style={{ width: '95%'}}>
            <List users={ users.list }></List>
        </View>
    )
    
}

export default UserList;