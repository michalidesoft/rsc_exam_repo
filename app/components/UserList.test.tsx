import React from 'react';
import { Provider } from 'react-redux';
import { createStore, combineReducers } from 'redux';
import renderer from 'react-test-renderer';
import { enableFetchMocks } from 'jest-fetch-mock'

import UserList from './UserList';
import usersReducer from '../reducers/usersReducer';


/** avoid async calls to the backend API in useEffect */
enableFetchMocks()

const rootReducer = combineReducers({ 
    users: usersReducer 
});

describe('<UserList />', () => {
    it('renders Loading...', () => {
        const mockStore = createStore(rootReducer, { users: {
            list: null,
            loading: true,
            error: null
        }});

        const tree = renderer.create(
            <Provider store={mockStore}>
                <UserList></UserList>
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});