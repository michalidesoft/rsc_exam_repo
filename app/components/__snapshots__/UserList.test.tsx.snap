// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserList /> renders Loading... 1`] = `
<View>
  <Text>
    Loading...
  </Text>
</View>
`;
